### DISCUSSION TITLE
Proposal: Infer return type of expression bodied local functions

### DISCUSSION THREAD

**Discussion by gulshan** [dotnet/csharplang#265] [General]
When local functions were first proposed, the return type was supposed to be inferred. But the idea was complex to implement and was postponed. Expression bodied local functions will be the simplest to infer the return type. Please do this in some upcoming 7.x release. Local functions with single return statements can also be considered for inferring. Example (not that good)-

```csharp
int add(int a, int b) => a+b;
// can be written as
var add(int a, int b) => a+b;
````

ref: https://github.com/dotnet/roslyn/issues/2930#issuecomment-276115005

**Comment by gafter** [comment]
Now that local functions can be recursive and can forward-reference other local functions, supporting inference in this way isn't a cleanly defined problem. I don't believe we're ever likely to do it.

**Comment by HaloFour** [comment]
@gafter 

Could the compiler be made to check for those situations and allow inference of the return value where the local function isn't recursive or invoke other local functions from which these complications would arise?

https://github.com/dotnet/csharplang/issues/249#issuecomment-286750761

**Comment by jnm2** [comment]
`var` for return type feels so wrong but it can't be helped I guess. `auto` would make sense but `var` has always meant "create a local variable."

**Comment by chergey** [comment]
@jnm2 `auto` may be unavailable (i.e. break present codebase)

**Comment by gafter** [comment]
@HaloFour 
> Could the compiler be made to check for those situations and allow inference of the return value where the local function isn't recursive or invoke other local functions from which these complications would arise?

No, not without that being described in the language specification somehow. I don't think there is a clean way to do that.

**Comment by bondsbw** [comment]
For that matter, would it be possible to infer the entire signature, given that inference only needs to search locally?

```c#
var add(a, b) => a+b;
var x = add(1, 5);
```

**Comment by bondsbw** [comment]
And then why not allow in the same block:

```c#
var y = add(1.5, 2.25);
```

It would generate two overloads for `add`, one for `(int, int) => int` and another `(double, double) => double`.

**Comment by onionhammer** [comment]
how about `let value (int x, int y) => x * y` ;P

**Comment by lucasteles** [comment]
@onionhammer i like it,  'let' is being considered for local immutable references to, at #421 and #188
